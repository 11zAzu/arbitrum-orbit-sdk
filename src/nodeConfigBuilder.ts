import { NodeConfig, NodeConfigOption } from './types/NodeConfig.generated';
import { nodeConfigBuilderDefaults } from './nodeConfigBuilderDefaults';

// todo: is there a way to make jsdoc readable when working with the builder?

type NodeConfigOptionName = NodeConfigOption['name'];

type NodeConfigOptionValue<TName extends NodeConfigOptionName> = Extract<
  NodeConfigOption,
  { name: TName }
>['type'];

const config = {
  auth: {
    /**
     * cool cool
     */
    api: 'auth.api',
    /**
     * asdf asdf
     */
    addr: 'auth.addr',
  },
} as const;

class NodeConfigBuilder {
  private nodeConfig: NodeConfig;

  constructor(initialNodeConfig?: NodeConfig) {
    this.nodeConfig = initialNodeConfig ?? {};
  }

  set<TName extends NodeConfigOptionName>(
    name: TName,
    value: NodeConfigOptionValue<TName>,
  ): NodeConfigBuilder {
    return this;
  }

  build(): NodeConfig {
    return this.nodeConfig;
  }
}

type CreateNodeConfigBuilderParams = {
  withoutDefaults: boolean;
};

export function createNodeConfigBuilder(params?: CreateNodeConfigBuilderParams): NodeConfigBuilder {
  const withoutDefaults = params?.withoutDefaults ?? false;

  if (withoutDefaults) {
    return new NodeConfigBuilder();
  }

  return new NodeConfigBuilder(nodeConfigBuilderDefaults);
}
