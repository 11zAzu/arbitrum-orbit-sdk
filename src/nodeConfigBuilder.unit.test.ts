import { it, expect } from 'vitest';

import { createNodeConfigBuilder } from './nodeConfigBuilder';

it('creates node config with defaults', () => {
  const nodeConfig = createNodeConfigBuilder().build();
  expect(nodeConfig).toMatchSnapshot();
});

it('creates node config without defaults', () => {
  const nodeConfig = createNodeConfigBuilder({ withoutDefaults: true }).build();
  expect(nodeConfig).toMatchSnapshot();
});
