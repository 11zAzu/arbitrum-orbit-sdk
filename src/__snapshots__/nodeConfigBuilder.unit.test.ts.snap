// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates node config with defaults 1`] = `
{
  "execution": {
    "caching": {
      "archive": true,
    },
    "forwarding-target": "",
    "sequencer": {
      "enable": true,
      "max-block-speed": "250ms",
      "max-tx-data-size": 85000,
    },
  },
  "http": {
    "addr": "0.0.0.0",
    "api": [
      "eth",
      "net",
      "web3",
      "arb",
      "debug",
    ],
    "corsdomain": [
      "*",
    ],
    "port": 8449,
    "vhosts": [
      "*",
    ],
  },
  "node": {
    "batch-poster": {
      "enable": true,
      "max-size": 90000,
    },
    "dangerous": {
      "no-sequencer-coordinator": true,
    },
    "delayed-sequencer": {
      "enable": true,
      "finalize-distance": 1,
      "use-merge-finality": false,
    },
    "sequencer": true,
    "staker": {
      "enable": true,
      "strategy": "MakeNodes",
    },
  },
}
`;

exports[`creates node config without defaults 1`] = `{}`;
